provisioning/device/src/Transports/Amqp/AmqpClientConnection.cs

```diff
@@ -34,121 +38,149 @@ internal AmqpClientConnection(Uri uri, AmqpSettings amqpSettings)
+                        _tcs = new TaskCompletionSource<TransportBase>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                        var args = new TransportAsyncCallbackArgs();
+                        args.SetBuffer(buffer.Buffer, buffer.Offset, buffer.Length);
+                        args.CompletedCallback = OnWriteHeaderComplete;
+                        args.Transport = _transport;
+                        bool operationPending = _transport.WriteAsync(args);
+
+                        if (Logging.IsEnabled)
+                            Logging.Info(
+                                this,
+                                $"{nameof(AmqpClientConnection)}.{nameof(OpenAsync)}: " +
+                                $"Sent Protocol Header: {_sentHeader} operationPending: {operationPending} completedSynchronously: {args.CompletedSynchronously}");
+
+                        if (!operationPending)
+                        {
+                            args.CompletedCallback(args);
+                        }
+
+                        _transport = await _tcs.Task.ConfigureAwait(false);
+                        await _transport.OpenAsync(cancellationToken).ConfigureAwait(false);
                     }
-
-                    _transport = await _tcs.Task.ConfigureAwait(false);
-                    await _transport.OpenAsync(cancellationToken).ConfigureAwait(false);
                 }
+                else
+                {
+                    var tcpInitiator = new AmqpTransportInitiator(_amqpSettings, TransportSettings);
+                    _transport = await tcpInitiator.ConnectAsync(cancellationToken).ConfigureAwait(false);
+                }
+
+                AmqpConnection = new AmqpConnection(_transport, _amqpSettings, AmqpConnectionSettings);
+                AmqpConnection.Closed += OnConnectionClosed;
+                await AmqpConnection.OpenAsync(cancellationToken).ConfigureAwait(false);
+                IsConnectionClosed = false;
             }
-            else
+            finally
             {
-                var tcpInitiator = new AmqpTransportInitiator(_amqpSettings, TransportSettings);
-                _transport = await tcpInitiator.ConnectAsync(cancellationToken).ConfigureAwait(false);
-            }
+                _connectionSemaphore.Release();
 
-            AmqpConnection = new AmqpConnection(_transport, _amqpSettings, AmqpConnectionSettings);
-            AmqpConnection.Closed += OnConnectionClosed;
-            await AmqpConnection.OpenAsync(cancellationToken).ConfigureAwait(false);
-            IsConnectionClosed = false;
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, $"{nameof(AmqpClientConnection)}.{nameof(OpenAsync)}");
+            }
         }
 
-        public async Task CloseAsync(CancellationToken cancellationToken)
+        internal async Task CloseAsync(CancellationToken cancellationToken)
         {
-            AmqpConnection connection = AmqpConnection;
```

Comment:
I don't understand this logic. Why get the property and then close, but not set it as null after?

Reply:
What you've changed it to makes more sense to me than it did previously. I think it was really just trying to not call close if it was already null.

Reply:
+1